src/arcsf/data/data_module.py

```diff
@@ -0,0 +1,207 @@
+
+        split_dict = {
+            "retain": retain_data,
+            "forget": forget_data,
+            "all": ConcatDataset([retain_data, forget_data]),
+        }
+        self.data = split_dict[split]
+
+    def __len__(self):
+        return len(self.data)
+
+    def __getitem__(self, idx):
+
+        question = self.data[idx]["question"]
+        answer = self.qa_formatter(self.data[idx]["answer"])
+
+        inp = self.qa_formatter((question, answer))
+
+        return self.tokenizer(inp)
+
+
+class QAForgetDataSet(Dataset):
+    """
+    Q+A Forget version of the dataset, __getitem__ returns a retain and forget sample.
+    Both are formatted using a question formatter. There is an option to output samples
+    using "I don't know" synonyms by specifying loss_type as "idk".
+    """
+
+    def __init__(
+        self,
+        data,
+        tokenizer,
+        qa_formatter,
+        loss_type,
+        random_seed=42,
+    ):
+        super().__init__()
+        self.tokenizer = tokenizer
+        self.qa_formatter = qa_formatter
+        self.loss_type = loss_type
+
+        self.forget_data, self.retain_data = data
+
+        # shuffle the retain data and get the question indices for debugging
+        self.retain_data = self.retain_data.shuffle(seed=random_seed)
+        self.retain_permutation = self.retain_data["question_index"]
+        # set item index acting as a counter for retain permutation
+        self.item_index = 0
+        self.retain_length = len(self.retain_data)
+
+        if loss_type == "idk":
+            with open("src/arcsf/data/idk.jsonl") as idk_file:
+                self.idk = idk_file.read().splitlines()
+            self.answer_sampler = self.idk_sampler
+        else:
+            self.answer_sampler = self.forget_sampler
+
+    def idk_sampler(self, _):
+        rand_pos = torch.randint(0, len(self.idk), (1,)).item()
+        return self.idk[rand_pos]
+
+    def forget_sampler(self, forget_row):
+        return forget_row["answer"]
+
+    def __len__(self):
+        return len(self.data)
+
+    def __getitem__(self, idx):
+        # this takes the first item in our retain data permutation using item_index
+        retain_row = self.retain_data[self.item_index]
+        # then rolls the permutation vector using the item_index to ensure
+        # samples aren't reused without first exhausting all retain samples
+        self.item_index = (self.item_index + 1) % self.retain_length
+
+        retain_question = retain_row["question"]
+        retain_answer = retain_row["answer"]
+
+        # forget data works as in above examples
+        forget_row = self.forget_data[idx]
+        forget_question = forget_row["question"]
+
+        forget_answer = self.answer_sampler(forget_row)
+
+        retain = self.qa_formatter((retain_question, retain_answer))
+        forget = self.qa_formatter((forget_question, forget_answer))
+
+        return self.tokenizer(retain), self.tokenizer(forget)
```

Comment:
There will likely need to be more logic than only calling the tokenizer here and in the other classes, e.g. see `convert_raw_data_to_model_format` in the TOFU code, but it may be easier to figure out when integrating this with the fine-tunung/forgetting code.

Reply:
Yeah I thought so, this was intended as more of a placeholder, until we know what we want it to do. I thought qa_formatter may be able to handle this logic? as in the TOFU code.